### PR TITLE
NetSim: Fix excessive wire deletes

### PR DESCRIPTION
**Pull request by islemaster**
This is a bugfix and general cleanup of the kick/reconnect process for idle clients.

It resolves:
- [#92868056 "Track down excessive wire deletes."](https://www.pivotaltracker.com/story/show/92868056)
- [#91731912 "On connection failure or timeout, visualization stays in background"](https://www.pivotaltracker.com/story/show/91731912)

It is also progress toward solving:
- [#92717990 "Router Connection Issues"](https://www.pivotaltracker.com/story/show/92717990)
- [#92718598 "To Investigate: Copies of people, bad extra nodes, cleanup not working"](https://www.pivotaltracker.com/story/show/92718598)

Understanding these changes will require some context.

**The problem**
The issue came to our attention through DynamoDB alarms indicating that our write capacity was being exceeded.  These alarms were suspicious for three reasons:
1. Only a few people (no more than ten) were using NetSim internally, and then not using it heavily.
2. We were nowhere near our read capacity, and NetSim should generate roughly equal read and write traffic.
3. Close examination showed that all throttled requests were deletes on the wire table, which should be significantly less write-active than the heartbeat, message or log tables.

This raised the possibility of some recursive wire-delete loop allowing a single client to generate enough `DELETE` requests to reach our (currently low) service capacity.  It took a couple of weeks, but I finally caught this loop in action on my local machine - it only happens when the client has been kicked out of the simulation by another client.

**The cleanup system**
Because there is no server component to NetSim except for shared storage, maintenance and cleanup of that shared storage is managed collectively by the connected clients.  Each client will periodically run a "cleaning routine" where it looks for expired rows on its shard and deletes them.

To keep from being deleted by its peers, each client maintains a 'heartbeat' row that it updates every six seconds with a current timestamp as a proxy for its 'presence' on the shard.  If a client goes inactive and its heartbeat is not updated for a full minute, another client may clean that node up.

Unfortunately, it's easy for a client to be idle for a minute, even if the page is open, because switching to another tab interferes with the heartbeat signal.  In the future I should [use a HTML5 Web Worker](http://stackoverflow.com/questions/5927284/how-can-i-make-setinterval-also-work-when-a-tab-is-inactive-in-chrome/5927432#12522580) to keep these presence messages alive when backgrounded, but that's another task.

**The root cause**
When a client was refocused and detected that its heartbeat was gone from storage, it would attempt its own full disconnect process.  I wrote this disconnect process doing what I _thought_ was right and backing out of it if an error occurred during disconnect.

Unfortunately in the case where we disconnect in response to someone else's cleaning routine, the first step in the disconnect process (removing our wire) would fail with a 404 and leave the local client thinking it was connected.  Then a moment later it would detect that its heartbeat was missing and start the process again, resulting in a neverending loop of wire `DELETE` calls.

**The solution**
When an error happens during disconnect, we continue disconnecting anyway.  At some level we can only make a "good faith" effort to clean up after ourselves and let the real cleanup system take care of it if something goes wrong - it's better than locking the local client into a bad state and overloading DynamoDB.

After correcting this with the wire deletes I found a similar problem a layer up with the node deletes as well.  I wrote in some debug buttons to make it easy to reproduce the problem and tuned things until the client can gracefully detect that it's been disconnected and reconnect itself to the shard.

auto-reconnect-on-kick

In this animation the client on the right kicks the client on the left out.  A few seconds later the client on the left detects this and jumps back to the lobby, reconnected.  Those big yellow buttons are the temporary debug buttons.  I'm hiding them in this PR but making it easy to re-enable them as it will be useful to test the "kicked/reconnect" flow in the future.

### CODE REVIEW

File: apps/src/netsim/NetSimLobby.js [hunk 1]
```diff
@@ -236,31 +236,45 @@ NetSimLobby.prototype.setShardID = function (shardID) {
...
-    this.shard_.nodeTable.tableChange.unregister(this.eventKeys.nodeTable);
-  }
-  if (this.eventKeys.wireTable) {
-    this.shard_.wireTable.tableChange.unregister(this.eventKeys.wireTable);
+  if (this.eventKeys.registeredShard) {
+    this.eventKeys.registeredShard.nodeTable.tableChange.unregister(
+        this.eventKeys.nodeTable);
+    this.eventKeys.registeredShard.wireTable.tableChange.unregister(
+        this.eventKeys.wireTable);
+    this.registeredShard = undefined;
```
Comment (islemaster) [inline apps/src/netsim/NetSimLobby.js]: Here I'm switching to a newer, safer pattern I've been using, where we actually cache off the thing we need to unregister from.

File: apps/src/netsim/NetSimLobby.js [hunk 2]
```diff
@@ -236,31 +236,45 @@ NetSimLobby.prototype.setShardID = function (shardID) {
...
   }
 
   this.shard_ = shard;
   this.myNode_ = myNode;
 
-  if (!this.shard_) {
-    // If we disconnected, just clear our lobby data
+  if (this.shard_) {
```
Comment (islemaster) [inline apps/src/netsim/NetSimLobby.js]: These two cases are complicated enough that I felt it was worth it to reverse their order and state things in the positive sense.

File: apps/src/netsim/NetSimLobby.js [hunk 3]
```diff
@@ -236,31 +236,45 @@ NetSimLobby.prototype.setShardID = function (shardID) {
...
-      this.onWireTableChange_.bind(this));
+    // Redraw the lobby
+    this.render();
 
-  // Trigger a forced read of the node table
-  this.fetchInitialLobbyData_();
+    // If there's only one option, try to auto-reconnect
+    if (this.shardChoices_.length === 1) {
+      this.setShardID(this.shardChoices_[0].shardID);
+    }
```
Comment (islemaster) [inline apps/src/netsim/NetSimLobby.js]: Three new things the lobby does when it detects a disconnect:
1. It clears the selectedShardID_ (we've been kicked, we don't have one now)
2. It triggers a `render()`; in the connect path this happens as a result of `fetchInitialLobbyData_()` but it was getting missed on disconnect
3. If there's only one possible shard anyway, trigger an auto-reconnect to that shard (which puts the student back in the "Pick a Peer/Router" screen)

File: apps/src/netsim/NetSimLocalClientNode.js [hunk 4]
```diff
@@ -412,19 +412,21 @@ NetSimLocalClientNode.prototype.synchronousDisconnectRemote = function () {
...
   this.myWire.destroy(function (err) {
+    // We're not going to stop if an error occurred here; the error might
+    // just be that the wire was already cleaned up by another node.
+    // As long as we make a good-faith disconnect effort, the cleanup system
+    // will correct any mistakes and we won't lock up our client trying to
+    // re-disconnect.
     if (err) {
-      onComplete(err);
-      return;
+      logger.info("Error while disconnecting: " + err.message);
```
Comment (islemaster) [inline apps/src/netsim/NetSimLocalClientNode.js]: Fix for the original problem with tons of wire deletes, and a long comment explaining why we don't return on error anymore here.

File: apps/src/netsim/NetSimVisualization.js [hunk 5]
```diff
@@ -192,6 +198,7 @@ NetSimVisualization.prototype.setLocalNode = function (newLocalNode) {
     this.localNode.isLocalNode = true;
   } else {
     this.localNode.kill();
+    this.localNode = null;
```
Comment (islemaster) [inline apps/src/netsim/NetSimVisualization.js]: This fixes a visualization bug that happened when a client was kicked - once you've called kill() on a viznode it's effectively invalidated and shouldn't be interacted with anymore (even though it takes a second to animate out), but we were hanging onto a reference to it and trying to update it when the client reconnected.  In fact, we want to drop our reference to this node so we make a whole new one on reconnect.

File: apps/src/netsim/netsim.js [hunk 6]
```diff
@@ -507,8 +513,10 @@ NetSim.prototype.disconnectFromShard = function (onComplete) {
   this.myNode.stopSimulation();
   this.myNode.destroy(function (err, result) {
     if (err) {
-      onComplete(err, result);
-      return;
+      logger.warn('Error destroying node:' + err.message);
+      // Don't stop disconnecting on an error here; we make a good-faith
+      // effort to clean up after ourselves, and let the cleaning system take
+      // care of the rest.
```
Comment (islemaster) [inline apps/src/netsim/netsim.js]: This is the second layer of the problem that I found after fixing the wire delete issue; same problem existed when trying to destroy the user node.

**Comment by davidsbailey** [comment]
Hi Brad, can you explain the motivation for the cleaning system? We pay for DynamoDB based on reserved throughput -- not bytes stored, not throughput used. Having clients do background access to clean up data will probably cause us to increase our reserved throughput since it will happen most during peak times. Alternatives include: (1) never clean anything up, or (2) clean up using a background process at non-peak times.

**Comment by islemaster** [comment]
Sure Dave.

The cleaning system went in before we knew we'd be backing NetSim with DynamoDB, so it's probably worth revisiting and optimizing to minimize throughput.  It serves a few purposes.

1. Nothing else is cleaning up very old records right now.  This should definitely be taken care of by a background process at some point, ideally with a single query (delete all netsim records older than a day) rather than a request per row.
2. A forcefully disconnected client (network outage, browser crash, etc) will stick around in the simulation until another client cleans them up.  This could be replaced by having clients simply _ignore_ expired records rather than actually deleting them, but besides complicating things a little it will increase table sizes.
3. Clients will frequently read the entire contents of a table (for their shard) in order to model the whole simulation locally, and because there is no "where clause" capability in the tables API right now.  Thus, keeping table sizes down while the simulation is in use helps keep our read throughput in check. There's already a potential leak here with router logs that grow indefinitely within a single session.  This could be solved by extending our API so that our tables can grow until off-peak cleanup without having an ever-increasing read cost.

That said, I think this hardening of the kick-recovery code is helpful in the short term to stabilize things for the ed team and stop the capacity alerts.  Optimizing for DynamoDB is a larger task that I think we may not get to until late May/early June.

File: apps/src/netsim/netsim.js (hunk 6 above)
Comment (davidsbailey) [inline apps/src/netsim/netsim.js]: Good catch :)

File: apps/style/netsim/style.scss [hunk 7]
```diff
@@ -776,5 +776,20 @@ $wide-column-width: 121px;
       padding-right: 14px;
     }
   }
+
+  // Really garish terrible style for debug buttons that we should not ship.
```
Comment (davidsbailey) [inline apps/style/netsim/style.scss]: These styles are fine for this purpose. But, there are some debug controls which hover in the bottom right like the ones visible on this level (if logged in as admin) which you could copy: http://levelbuilder.studio.code.org/s/artist/stage/1/puzzle/10 
screen shot 2015-04-29 at 2 50 08 pm

**Comment by davidsbailey** [comment]
Insidious! Nice work tracking this (these!) down. LGTM.

**Comment by islemaster** [comment]
Thanks Dave!  Tracking optimization for DynamoDB here: https://www.pivotaltracker.com/story/show/93585704

I'll probably consider moving the debug to a floating element like the admin panel next time I need to add any.

**Pull request merged**